Allow for NBATH=0 and Fix Pair and Exc Susceptibilities

lcrippa opened the pull request:
For isolated impurity problems. Closes https://github.com/EDIpack/EDIpack/issues/22

aamaricci committed:
5.1.0 Updated test. Fixed bug in parse_input of array variables.
This version needs the latest SciFortran (32e597b865584b4 from LC 6/6/25)

lcrippa committed:
Merge branch 'master' of github.com:EDIpack/EDIpack

lcrippa committed:
Merge branch 'master' of github.com:EDIpack/EDIpack

lcrippa committed:
Allow for NBATH=0. The code just solves an isolated impurity

lcrippa committed:
r-DMFT and c-binding for Nbath=0

lcrippa committed:
Better print

lcrippa committed:
Return zero delta if no bath

lcrippa committed:
Added flags for nobath main functions

lcrippa committed:
Fixed NBATH=0 with debug flags

Fortran lets us allocate 0-dimensional arrays, but not access them.
So, we need to allocate the bath even if it is 0-d, but not access it.
For loops with upper bound 0 take care of themselves.

lcrippa committed:
Amendment to previous commit

lcrippa committed:
also write_dmft_bath needs not return anything

lcrippa commented:
Idea:

- fortran lets us allocate zero-dimensional arrays
- if a loop has upper bound lower than the counter, it is not executed

So: we still allocate the bath even if Nbath=0, and deallocate it at the end. We don't access it because all access is behind loops at H creation time. The routines that access it (fit, write_dmft_bath, read_dmft_bath) get a preliminary check and exit.
Delta returns an array of zeros.

lcrippa committed:
also set and get bath can do the preliminary check. Only allocate/deallocate have to run always

lcrippa committed:
Fix pair susceptibility

- first: only look for ksector if jsector is not 0 (makes sense since
later we check that they are both non-zero...)
- second: if one of those two sectors are zero, we still need to alloate
a GFmatrix but with Nexc=0. This was fixed in GF a long time ago

lcrippa committed:
Fixed exc susceptibility [skip ci]

This was more complicated. I did the following things

- I don't know why the gfmatrix was initialized with indices [0:2].
I get the singlet is different, but that doesn't work: when
write_gfmatrix is called, it expects indexing starting from 1.
Also, here xy are together. I don't understand clearly why, since we
always talk about 4 components. Anyway, I have corrected the indexing
and now it starts from 1.

- Same fix as before regarding the situation where add_to_lanczos is not
  called: i needed to add two logical variables (success_up, success_dw)
  and multiple calls to allocate_GFmatrix with Nexc=0 to make this work.

- In add_to_lanczos there was a further check of vnorm/=0. If that was
  the case it returned. But we need to call allocate_GFmatrix with
  Nexc=0 also there!

Now the code runs also for Nbath=0. I fixed the crash, but I'm not
totally confident the physics makes sense, so this needs review.

aamaricci committed:
Intermediate commit

aamaricci committed:
Merge branch 'zerobath' of github.com:EDIpack/EDIpack2.0 into zerobath

aamaricci reviewed (approved):
All changes are ok for me.
Nbath==0 is operative now.
all Chi functions have been checked and fixed.

Run tests for normal/normal with all chi_flags=T and Nbath=0,1,2,3,4 in serial and mpi (np=2,4) in DEBUG mode. No running issues so far.

aamaricci committed:
Update
Changed a bit the fixings introduced by LC to make the ED_CHI_? working.
Just a little simplification and re-organization to make code more homogeneous.

Ready to merge into master.

lcrippa committed:
Update workflows

tests: To run on the branch which triggered it
docs:  To only run on pushes to master

lcrippa committed:
CI: avoid duplicated runs

lcrippa committed:
CI: fix branch name

lcrippa committed:
Added explanation for future memory [skip ci]

aamaricci committed:
Fixed documentation for ED_CHI_EXCT
